### PR TITLE
[libc] Add C23 limits.h header.

### PR DESCRIPTION
**Pull request by lntue**

### CODE REVIEW

File: libc/include/llvm-libc-macros/limits-macros.h [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __LLVM_LIBC_MACROS_LIMITS_MACROS_H
+#define __LLVM_LIBC_MACROS_LIMITS_MACROS_H
+
+#ifdef __clang__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wgnu-include-next"
```
Comment (jhuber6) [inline libc/include/llvm-libc-macros/limits-macros.h:14]: What warning is this hiding? Might be able to work around it with `__has_include_next` or something.

Comment (lntue) [reply libc/include/llvm-libc-macros/limits-macros.h:14]: It is to hide the following warning from clang:
```
limits-macros.h:17:2: warning: #include_next is a language extension [-Wgnu-include-next]
```
There are the same warnings from gcc, but I'm still tracking the correct `-W` flag to ignore.

File: libc/include/llvm-libc-macros/limits-macros.h [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+#else
+#define CHAR_BIT 8
+#endif // __CHAR_BIT__
+#endif // CHAR_BIT
+
+// TODO: define MB_LEN_MAX if missing
+//   clang: MB_LEN_MAX = 1 -
+//   https://github.com/llvm/llvm-project/blob/main/clang/lib/Headers/limits.h#L64
+//   glibc: MB_LEN_MAX = 16 -
+//   https://github.com/lattera/glibc/blob/master/include/limits.h#L32
```
Comment (SchrodingerZhu) [inline libc/include/llvm-libc-macros/limits-macros.h:35]: Maybe it is better to refer to `https://github.com/bminor/glibc/blob/master/include/limits.h`

File: libc/include/llvm-libc-macros/limits-macros.h [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+#define SCHAR_MAX 0x7f
+#endif // __SCHAR_MAX__
+#endif // SCHAR_MAX
+
+#ifndef UCHAR_MAX
+#define UCHAR_MAX (SCHAR_MAX * 2 + 1)
+#endif // UCHAR_MAX
+
+#ifndef CHAR_MAX
+#ifdef __CHAR_UNSIGNED__
```
Comment (SchrodingerZhu) [inline libc/include/llvm-libc-macros/limits-macros.h:124]: Maybe
```c
#ifndef __CHAR_UNSIGNED__
#define __CHAR_UNSIGNED__ ('\xff' > 0)
#endif
```

Comment (lntue) [reply libc/include/llvm-libc-macros/limits-macros.h:124]: Do you know any compiler failed to define `__CHAR_UNSIGNED__` on such platforms?  I would assume gcc and clang are correct with this flag.

**Comment by lntue** [comment]
@petrhosek Do you mind checking if this header is correct for arm32 / riscv32 / baremetal?  Can I leave it to you to enable this header for those targets later?  Thanks,

File: libc/include/llvm-libc-macros/limits-macros.h (hunk 3 above)
Comment (SchrodingerZhu) [reply libc/include/llvm-libc-macros/limits-macros.h:124]: Not rly. But MUSL seems to be using this as a portable implementation. see https://github.com/bminor/musl/blob/master/include/limits.h#L10

File: libc/include/llvm-libc-macros/limits-macros.h (hunk 2 above)
Comment (lntue) [reply libc/include/llvm-libc-macros/limits-macros.h]: Done

File: libc/include/llvm-libc-macros/limits-macros.h (hunk 3 above)
Comment (lntue) [reply libc/include/llvm-libc-macros/limits-macros.h:132]: Thanks, I've added extra check for char's signed-ness.

File: libc/include/llvm-libc-macros/limits-macros.h (hunk 1 above)
Comment (lntue) [reply libc/include/llvm-libc-macros/limits-macros.h:15]: `#pragma GCC system_header` does the trick for gcc.

**Review by SchrodingerZhu — APPROVED** [review]

File: libc/include/llvm-libc-macros/float-macros.h [hunk 4]
```diff
@@ -9,8 +9,20 @@
...
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wgnu-include-next"
+#else // gcc
+#pragma GCC system_header
+#endif //__clang__
+
 #include_next <float.h>
 
+#ifdef __clang__
+#pragma GCC diagnostic pop
```
Comment (nickdesaulniers) [inline libc/include/llvm-libc-macros/float-macros.h:23]: If you're only using these pragmas for clang, then please use `#pragma clang` rather than `#pragma GCC`.

File: libc/include/llvm-libc-macros/limits-macros.h (hunk 1 above)
Comment (nickdesaulniers) [reply libc/include/llvm-libc-macros/limits-macros.h:15]: This is because include_next wasn't standardized until C23.

**Comment by michaelrj-google** [comment]
Ideally we wouldn't provide this header, since it's provided by the compiler. I'm not sure this will fix the problem on the gcc bot since this still includes the system's header, which is what is missing the `__GLIBC_USE` macro.

**Review by nickdesaulniers — CHANGES_REQUESTED** [review]
Putting a hold on this to discuss.

AFAIK, the compiler should be providing these definitions.  Under what conditions is this header needed?

**Comment by nickdesaulniers** [comment]
```
/usr/include/limits.h:145:17: error: missing binary operator before token "("
```
on my system, that file and line 145 use `__GLIBC_USE`.  There's a comment block at the top of that file that says
```
The GNU C Library is distributed
```
and later has:
```
/* Get the compiler's limits.h, which defines almost all the ISO constants.
# include_next <limits.h>
```
so it seems what's happening is that we have code (somewhere) that is including `<limits.h>` and getting the glibc header.

Perhaps we need an `#include_next <limits.h>` under some very specific conditions.

**Comment by nickdesaulniers** [comment]
src/__support/CPP/limits.h is including limits.h and getting glibc's limits.h, not the compilers'.
```
In file included from /usr/local/lib/gcc/x86_64-pc-linux-gnu/12.2.0/include-fixed/limits.h:203,
                 from /usr/local/lib/gcc/x86_64-pc-linux-gnu/12.2.0/include-fixed/syslimits.h:7,
                 from /usr/local/lib/gcc/x86_64-pc-linux-gnu/12.2.0/include-fixed/limits.h:34,
-->              from /home/llvm-libc-buildbot/buildbot-worker/libc-x86_64-debian-fullbuild/libc-x86_64-debian-gcc-fullbuild-dbg/llvm-project/libc/src/__support/CPP/limits.h:16,
                 from /home/llvm-libc-buildbot/buildbot-worker/libc-x86_64-debian-fullbuild/libc-x86_64-debian-gcc-fullbuild-dbg/llvm-project/libc/src/__support/str_to_integer.h:12,
                 from /home/llvm-libc-buildbot/buildbot-worker/libc-x86_64-debian-fullbuild/libc-x86_64-debian-gcc-fullbuild-dbg/llvm-project/libc/src/inttypes/strtoumax.cpp:11:
/usr/include/limits.h:145:17: error: missing binary operator before token "("
  145 | #if __GLIBC_USE (IEC_605
```

File: libc/include/llvm-libc-macros/float-macros.h (hunk 4 above)
Comment (lntue) [reply libc/include/llvm-libc-macros/float-macros.h]: Done.

**Comment by nickdesaulniers** [comment]
> recent GCC does not seem to provide a standalone version of limits.h on my PC:
> lib/gcc/x86_64-pc-linux-gnu/13.2.1/include/limits.h

^ it's right there. `lib/gcc/x86_64-pc-linux-gnu/13.2.1/include/` is in the compiler's default include path.

**Comment by SchrodingerZhu** [comment]
I see.

File: libc/include/llvm-libc-macros/limits-macros.h [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+#if __has_include_next(<limits.h>)
+
+// Suppress `#include_next is a language extension` warnings.
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wgnu-include-next"
+#elif defined(__GNUC__) // gcc
+#if !defined _GCC_LIMITS_H_
+#define _GCC_LIMITS_H_
```
Comment (nickdesaulniers) [inline libc/include/llvm-libc-macros/limits-macros.h:20]: Needs a comment explaining what's going on with GCC's limits.h trying to include_next limits.h.

File: libc/include/llvm-libc-macros/limits-macros.h [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+#define __LLVM_LIBC_MACROS_LIMITS_MACROS_H
+
+#if __has_include_next(<limits.h>)
+
+// Suppress `#include_next is a language extension` warnings.
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wgnu-include-next"
+#elif defined(__GNUC__) // gcc
+#if !defined _GCC_LIMITS_H_
```
Comment (nickdesaulniers) [inline libc/include/llvm-libc-macros/limits-macros.h:19]: These two can be combined.

```c
#elif defined(__GNUC__) && !defined(_GCC_LIMITS_H_)
```

Make sure to be consistent in:
1. your use of `ifdef` for simple macros, vs `if defined` when there's multiple things to check.
2. your use of `()` around the token being checked.

File: libc/include/llvm-libc-macros/limits-macros.h [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+#include_next <limits.h>
+
+#ifdef __clang__
+#pragma clang diagnostic pop
+#endif // __clang__
+
+#endif // __has_include_next(<limits.h>)
+
+// Supplement missing macros.
+
```
Comment (nickdesaulniers) [inline libc/include/llvm-libc-macros/limits-macros.h:35]: For which compiler/version/env are these macros missing? (Especially since I'd expect these to be provided by the compiler).

Would be good to comment in the source.  I'd like to triple check that these are indeed missing in those environments.

File: libc/include/llvm-libc-macros/limits-macros.h (hunk 5 above)
Comment (nickdesaulniers) [reply libc/include/llvm-libc-macros/limits-macros.h:20]: It might be cleaner to move this into its own guard with check, rather than keeping in with the pragmas.

Example:
```c
#if defined(__GNUC__) && !defined(__clang__) && !defined(_GCC_LIMITS_H_)
// comment why
#define _GCC_LIMITS_H_
#endif

#ifdef __clang__
#pragma clang diagnostic push
...
```

File: libc/include/llvm-libc-macros/limits-macros.h (hunk 6 above)
Comment (lntue) [reply libc/include/llvm-libc-macros/limits-macros.h]: Done.

File: libc/include/llvm-libc-macros/limits-macros.h (hunk 5 above)
Comment (lntue) [reply libc/include/llvm-libc-macros/limits-macros.h]: Done.

File: libc/include/llvm-libc-macros/limits-macros.h [hunk 8]
```diff
@@ -15,13 +15,16 @@
...
 #pragma clang diagnostic ignored "-Wgnu-include-next"
-#elif defined(__GNUC__) // gcc
-#if !defined _GCC_LIMITS_H_
-#define _GCC_LIMITS_H_
-#endif
+#else // gcc
 #pragma GCC system_header
 #endif // __clang__, __GNUC__
 
+#if defined(__GNUC__) && !defined(_GCC_LIMITS_H_)
```
Comment (nickdesaulniers) [inline libc/include/llvm-libc-macros/limits-macros.h:22]: careful, clang's preprocessor also defines `__GNUC__` "for compatibility."

We only need to #define this for GCC due to GCC's limits.h (specific to GCC's resource dir, which is unused when CC=clang).

Comment (lntue) [reply libc/include/llvm-libc-macros/limits-macros.h]: Updated.

File: libc/include/llvm-libc-macros/limits-macros.h (hunk 7 above)
Comment (lntue) [reply libc/include/llvm-libc-macros/limits-macros.h]: Besides some old versions of gcc do not have `LLONG_MIN`, `LLONG_MAX`, `ULLONG_MAX` as in glibc's limits.h comments, there are also C23 new macros `*_WIDTH` that are only added to both clang and gcc recently.  The rest are for completeness.

Comment (nickdesaulniers) [reply libc/include/llvm-libc-macros/limits-macros.h]: > Besides some old versions of gcc do not have LLONG_MIN, LLONG_MAX, ULLONG_MAX as in glibc's limits.h comments

Ah, but do we have to support any of those versions of GCC?  If not, I don't think we should add these.

> there are also C23 new macros *_WIDTH that are only added to both clang and gcc recently.

Ok, those are fair to retain with a comment //why//.

> The rest are for completeness.

Right, but we don't //need// them; they should come from the compiler's resource headers which we should always expect to be able to rely on.  If they don't, then I think it would be better to:
1. allow the build to fail loudly because such preprocessor defines are missing.
2. add the bare minimum to get the build back to green, with comments/commit messages explaining why.

The thing I worry most about overriding the compiler's resource headers is that we risk getting these wrong, which may lead to hard to diagnose issues, particularly on platforms other than the host machine.

File: libc/include/llvm-libc-macros/limits-macros.h [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wgnu-include-next"
+#else // gcc
+#pragma GCC system_header
+#endif // __clang__, __GNUC__
+
+#if defined(__GNUC__) && !defined(__clang__) && !defined(_GCC_LIMITS_H_)
+// The system's limits.h may, in turn, try to #include_next GCC's limits.h when
+// building with GCC.
+// Define the macro _GCC_LIMITS_H_ to stop its chains of #include_next.
```
Comment (nickdesaulniers) [inline libc/include/llvm-libc-macros/limits-macros.h:25]: It's the other way around.  It's gcc's limits.h which includes gcc's syslimits.h which reincluded gcc's limits.h which finally include_next's glibc's limits.h which is the problem.  I would reword this as:

> Prevent GCC from including glibc's limits.h.

---

For reference the include chain looks like:
```
In file included:
(GLIBC) from /usr/include/limits.h:26,
(GCC) from /usr/lib/gcc/x86_64-linux-gnu/13/include/limits.h:205,
(GCC) from /usr/lib/gcc/x86_64-linux-gnu/13/include/syslimits.h:7,
(GCC) from /usr/lib/gcc/x86_64-linux-gnu/13/include/limits.h:34,
(LLVMLIBC) from /android0/llvm-project/libc/src/__support/CPP/limits.h:16,
```

File: libc/include/llvm-libc-macros/limits-macros.h [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+
+#ifndef __LLVM_LIBC_MACROS_LIMITS_MACROS_H
+#define __LLVM_LIBC_MACROS_LIMITS_MACROS_H
+
+#if __has_include_next(<limits.h>)
+
+// Suppress `#include_next is a language extension` warnings.
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wgnu-include-next"
```
Comment (nickdesaulniers) [inline libc/include/llvm-libc-macros/limits-macros.h:17]: If these pragmas keep popping up, we might want to consider making a dedicated macro for this that is reusable and more concise. `_Pragma` exists to have pragmas in preprocessor macros.  Not required for this PR, but food for thought in the future.

**Comment by nickdesaulniers** [comment]
FWIW, I was able to get the GCC build back to green with just:
```diff
diff --git a/libc/src/__support/CPP/limits.h b/libc/src/__support/CPP/limits.h
index d9e7090a0b7b..2cc76fe4bbf7 100644
--- a/libc/src/__support/CPP/limits.h
+++ b/libc/src/__support/CPP/limits.h
@@ -13,6 +13,9 @@
 #include "src/__support/CPP/type_traits/is_signed.h"
 #include "src/__support/macros/attributes.h" // LIBC_INLINE
 
+#if !defined(__clang__)
+#define _LIBC_LIMITS_H_
+#endif
 #include <limits.h> // CHAR_BIT
 
 namespace LIBC_NAMESPACE {
diff --git a/libc/src/time/mktime.cpp b/libc/src/time/mktime.cpp
index e57565f0d1fe..9ea92ebcde3c 100644
--- a/libc/src/time/mktime.cpp
+++ b/libc/src/time/mktime.cpp
@@ -10,6 +10,9 @@
 #include "src/__support/common.h"
 #include "src/time/time_utils.h"
 
+#if !defined(__clang__)
+#define _LIBC_LIMITS_H_
+#endif
 #include <limits.h>
 
 namespace LIBC_NAMESPACE {
diff --git a/libc/src/time/time_utils.cpp b/libc/src/time/time_utils.cpp
index 199a74cb168a..ccdd4d7de50c 100644
--- a/libc/src/time/time_utils.cpp
+++ b/libc/src/time/time_utils.cpp
@@ -9,6 +9,9 @@
 #include "src/time/time_utils.h"
 #include "src/__support/common.h"
 
+#if !defined(__clang__)
+#define _LIBC_LIMITS_H_
+#endif
 #include <limits.h>
 
 namespace LIBC_NAMESPACE {
```
I think we could provide our own limits.h (as you do here) and just do that one define if not clang; then require all TUs to include our limits.h, not `<limits.h>` since it's a tangly mess of include_next for different compiler's resource headers.

File: libc/include/llvm-libc-macros/limits-macros.h (hunk 7 above)
Comment (nickdesaulniers) [reply libc/include/llvm-libc-macros/limits-macros.h]: > Right, but we don't //need// them; they should come from the compiler's resource headers which we should always expect to be able to rely on.

Ok, so looking at GCC's resource headers, it seems that GCC only defines `LLONG_MIN`, `LLONG_MAX` and `ULLONG_MAX` `if defined (__STDC_VERSION__) && __STDC_VERSION__ >= 199901L` which I think is a problem for C++, and our libc/test/src/__support/CPP/limits_test.cpp which references those.  So those are the minimum we may want to just define ourselves.

File: libc/src/__support/CPP/limits.h [hunk 11]
```diff
@@ -9,23 +9,14 @@
...
 #include "src/__support/CPP/type_traits/is_integral.h"
 #include "src/__support/CPP/type_traits/is_signed.h"
 #include "src/__support/macros/attributes.h" // LIBC_INLINE
 
-#include <limits.h> // CHAR_BIT
-
 namespace LIBC_NAMESPACE {
 namespace cpp {
 
-// Some older gcc distributions don't define these for 32 bit targets.
```
Comment (nickdesaulniers) [inline libc/src/__support/CPP/limits.h:21]: Is there any information on when this was fixed, such as a specific version of GCC?

Especially since this is src/ so we don't need to support versions of GCC older than gcc-11.

I'd be curious how many of the fallbacks you could remove to satisfy the build with GCC; I suspect not all of the definitions are necessary.

File: libc/include/llvm-libc-macros/limits-macros.h [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+#endif // __clang__, __GNUC__
+
+#if defined(__GNUC__) && !defined(__clang__) && !defined(_GCC_LIMITS_H_)
+// The system's limits.h may, in turn, try to #include_next GCC's limits.h when
+// building with GCC.
+// Define the macro _GCC_LIMITS_H_ to stop its chains of #include_next.
+#define _GCC_LIMITS_H_
+#endif
+
+// Include compiler's header
```
Comment (nickdesaulniers) [inline libc/include/llvm-libc-macros/limits-macros.h:29]: Can you confirm that this is the compiler's header?

I guess I would have thought that `#include <limits.h>` _is_ the compiler's header, so `#include_next <limits.h>` is...perhaps `/usr/include/limits.h` which would be from glibc. The `-H` compiler flag can give you a very verbose output about the chain of includes.

File: libc/include/llvm-libc-macros/limits-macros.h (hunk 9 above)
Comment (nickdesaulniers) [reply libc/include/llvm-libc-macros/limits-macros.h:25]: Bumping this thread. I still think the comment need to be updated (or my understanding corrected).

File: libc/cmake/modules/LLVMLibCLibraryRules.cmake [hunk 13]
```diff
@@ -173,7 +173,17 @@ function(create_header_library fq_target_name)
...
-    target_link_libraries(${fq_target_name} INTERFACE ${ADD_HEADER_DEPENDS})
+
+    # `*.__copied_hdr__` is created only to copy the header files to the target
+    # location, not to be linked against.
+    set(link_lib "")
+    foreach(dep ${ADD_HEADER_DEPENDS})
+      if (NOT dep MATCHES "__copied_hdr__")
+        list(APPEND link_lib ${dep})
+      endif()
+    endforeach()
```
Comment (nickdesaulniers) [inline libc/cmake/modules/LLVMLibCLibraryRules.cmake:184]: Question: is this exposing your newly added limits.h to the consumers of llvm-libc?

Comment (lntue) [reply libc/cmake/modules/LLVMLibCLibraryRules.cmake:184]: No it's not.  It just remove those extra command-only targets.  Changes in this part  We won't add `limits.h` to the consumers, only the "internal" `include/llvm-libc-macros/limit-macros.h`.

File: libc/include/llvm-libc-macros/limits-macros.h (hunk 10 above)
Comment (lntue) [reply libc/include/llvm-libc-macros/limits-macros.h]: +1

**Review by nickdesaulniers — APPROVED** [review]

File: libc/include/llvm-libc-macros/limits-macros.h (hunk 9 above)
Comment (lntue) [reply libc/include/llvm-libc-macros/limits-macros.h]: Since the current state of those `limits.h` headers is messy, I just remove the `#include_next <limits.h>`, only use compiler supplied macros `__INT_MAX__` and friends, and filling in the rest of C23 `limits.h` macros if missing.

File: libc/src/__support/CPP/limits.h (hunk 11 above)
Comment (lntue) [reply libc/src/__support/CPP/limits.h:21]: I would defer this question to @michaelrj-google who encountered the issue.

File: libc/include/llvm-libc-macros/limits-macros.h (hunk 12 above)
Comment (lntue) [reply libc/include/llvm-libc-macros/limits-macros.h]: Drop `#include_next <limits.h>` to get out of these messy chains of header inclusion.
